File: typescript/day_6/part_2.ts
import { readFileSync } from 'fs';

interface Problem {
    numbers: number[];
    operation: string;
}

function solve(input: string): number {
    const lines = input.trim().split('\n');

    const operations_line = lines[lines.length - 1];
    const number_lines = lines.slice(0, lines.length - 1);

    let max_width = 0;
    for (const line of lines) {
        max_width = Math.max(max_width, line.length);
    }

    const padded_number_lines: string[] = [];
    for (const line of number_lines) {
        padded_number_lines.push(line.padEnd(max_width, ' '));
    }
    const padded_operations_line = operations_line.padEnd(max_width, ' ');

    const separator_columns: number[] = [];
    for (let col = 0; col < max_width; col++) {
        let is_separator = true;
        for (const line of padded_number_lines) {
            if (col < line.length && line[col] !== ' ') {
                is_separator = false;
                break;
            }
        }
        if (col < padded_operations_line.length && padded_operations_line[col] !== ' ') {
            is_separator = false;
        }

        if (is_separator) {
            separator_columns.push(col);
        }
    }

    const boundaries = [0, ...separator_columns, max_width];

    const problems: Problem[] = [];
    for (let i = 0; i < boundaries.length - 1; i++) {
        const start = boundaries[i];
        const end = boundaries[i + 1];

        let operation: string | null = null;
        for (let col = start; col < end; col++) {
            if (padded_operations_line[col] === '*' || padded_operations_line[col] === '+') {
                operation = padded_operations_line[col];
                break;
            }
        }

        if (operation === null) {
            continue;
        }

        const reversed_lines: string[] = [];
        for (const line of padded_number_lines) {
            const segment = line.substring(start, end);
            const reversed_segment = segment.split('').reverse().join('');
            reversed_lines.push(reversed_segment);
        }

        const numbers: number[] = [];
        const segment_width = end - start;

        for (let col = 0; col < segment_width; col++) {
            let digits = '';
            for (const reversed_line of reversed_lines) {
                if (col < reversed_line.length && reversed_line[col] !== ' ') {
                    digits += reversed_line[col];
                }
            }

            if (digits.length > 0) {
                const num = parseInt(digits);
                if (!isNaN(num)) {
                    numbers.push(num);
                }
            }
        }

        if (numbers.length > 0) {
            problems.push({ numbers, operation });
        }
    }

    let total = 0;

    for (const problem of problems) {
        if (problem.operation === '*') {
            let result = 1;
            for (const num of problem.numbers) {
                result *= num;
            }
            total += result;
        } else {
            let result = 0;
            for (const num of problem.numbers) {
                result += num;
            }
            total += result;
        }
    }

    return total;
}

const input = readFileSync('../inputs/day_6.txt', 'utf-8');
const result = solve(input);
console.log(result);

File: java/day_6/Part2.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class Part2 {
    static class Problem {
        List<Integer> numbers;
        char operation;
        Problem(List<Integer> numbers, char operation) {
            this.numbers = numbers;
            this.operation = operation;
        }
    }

    public static int solve(String input) {
        String[] lines = input.trim().split("\n");

        String operations_line = lines[lines.length - 1];
        List<String> number_lines = new ArrayList<>();
        for (int i = 0; i < lines.length - 1; i++) {
            number_lines.add(lines[i]);
        }

        int max_width = 0;
        for (String line : lines) {
            max_width = Math.max(max_width, line.length());
        }

        for (int i = 0; i < number_lines.size(); i++) {
            while (number_lines.get(i).length() < max_width) {
                number_lines.set(i, number_lines.get(i) + " ");
            }
        }
        while (operations_line.length() < max_width) {
            operations_line += " ";
        }

        List<Integer> separator_columns = new ArrayList<>();
        for (int col = 0; col < max_width; col++) {
            boolean is_separator = true;
            for (String line : number_lines) {
                if (col < line.length() && line.charAt(col) != ' ') {
                    is_separator = false;
                    break;
                }
            }
            if (col < operations_line.length() && operations_line.charAt(col) != ' ') {
                is_separator = false;
            }

            if (is_separator) {
                separator_columns.add(col);
            }
        }

        List<Integer> boundaries = new ArrayList<>();
        boundaries.add(0);
        boundaries.addAll(separator_columns);
        boundaries.add(max_width);

        List<Problem> problems = new ArrayList<>();
        for (int i = 0; i < boundaries.size() - 1; i++) {
            int start = boundaries.get(i);
            int end = boundaries.get(i + 1);

            Character operation = null;
            for (int col = start; col < end; col++) {
                if (operations_line.charAt(col) == '*' || operations_line.charAt(col) == '+') {
                    operation = operations_line.charAt(col);
                    break;
                }
            }

            if (operation == null) {
                continue;
            }

            List<String> reversed_lines = new ArrayList<>();
            for (String line : number_lines) {
                String segment = line.substring(start, end);
                String reversed_segment = new StringBuilder(segment).reverse().toString();
                reversed_lines.add(reversed_segment);
            }

            List<Integer> numbers = new ArrayList<>();
            int segment_width = end - start;

            for (int col = 0; col < segment_width; col++) {
                StringBuilder digits = new StringBuilder();
                for (String reversed_line : reversed_lines) {
                    if (col < reversed_line.length() && reversed_line.charAt(col) != ' ') {
                        digits.append(reversed_line.charAt(col));
                    }
                }

                if (digits.length() > 0) {
                    try {
                        numbers.add(Integer.parseInt(digits.toString()));
                    } catch (NumberFormatException e) {
                    }
                }
            }

            if (!numbers.isEmpty()) {
                problems.add(new Problem(numbers, operation));
            }
        }

        int total = 0;

        for (Problem problem : problems) {
            if (problem.operation == '*') {
                int result = 1;
                for (int num : problem.numbers) {
                    result *= num;
                }
                total += result;
            } else {
                int result = 0;
                for (int num : problem.numbers) {
                    result += num;
                }
                total += result;
            }
        }

        return total;
    }

    public static void main(String[] args) throws IOException {
        String input = new String(Files.readAllBytes(Paths.get("../inputs/day_6.txt")));
        int result = solve(input);
        System.out.println(result);
    }
}
